# **Heart Disease Prediction With Machine Learning**


# **1. Data Exploration**

## **Preparing the Environment**


To ensure reproducibility and isolate project dependencies, a dedicated Python virtual environment was utilized for this research. 
The environment was configured as follows:
* **Virtual Environment:** Initialized using `venv`.
* **Package Management:** `pip` upgraded to the latest version to prevent dependency resolution conflicts.
* **Dependencies:** Core libraries for data manipulation (`pandas`, `numpy`), visualization, and machine learning were successfully installed via `requirements.txt`.

### **Library Imports**

The following cell imports the fundamental libraries required for data ingestion and initial exploration.

In [1]:
# Importing Libraries

import pandas as pd          # For data manipulation
import numpy as np            # For mathematical operation
import matplotlib.pyplot as plt    # For data visualization
import seaborn as sns              # For data visualization

## **Data Ingestion**

In this section, the raw Framingham daatset is loaded into the working environment and copied into memory, thereby preparing it for exploration.



In [15]:
# Loading the Framingham dataset
framingham_df = pd.read_csv("../data/raw/framingham.csv")

df = framingham_df.copy()  # Creating a copy of the original dataset to work with   

print('Framingham Data Loaded Successfully: \n')

print("The first 5 rows of the Framingham data: \n")
df.head()

Framingham Data Loaded Successfully: 

The first 5 rows of the Framingham data: 



,male,age,education,currentSmoker,cigsPerDay,BPMeds,prevalentStroke,prevalentHyp,diabetes,totChol,sysBP,diaBP,BMI,heartRate,glucose,TenYearCHD
0,1,39,4.0,0,0.0,0.0,0,0,0,195.0,106.0,70.0,26.97,80.0,77.0,0
1,0,46,2.0,0,0.0,0.0,0,0,0,250.0,121.0,81.0,28.73,95.0,76.0,0
2,1,48,1.0,1,20.0,0.0,0,0,0,245.0,127.5,80.0,25.34,75.0,70.0,0
3,0,61,3.0,1,30.0,0.0,0,1,0,225.0,150.0,95.0,28.58,65.0,103.0,1
4,0,46,3.0,1,23.0,0.0,0,0,0,285.0,130.0,84.0,23.10,85.0,85.0,0


## **Data Overview**


This section provides an initial assessment of the dataset to understand its structure, completeness, and potential issues before cleaning.



In [16]:
# Returning the basic information about the data
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4240 entries, 0 to 4239
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4240 non-null   int64  
 1   age              4240 non-null   int64  
 2   education        4135 non-null   float64
 3   currentSmoker    4240 non-null   int64  
 4   cigsPerDay       4211 non-null   float64
 5   BPMeds           4187 non-null   float64
 6   prevalentStroke  4240 non-null   int64  
 7   prevalentHyp     4240 non-null   int64  
 8   diabetes         4240 non-null   int64  
 9   totChol          4190 non-null   float64
 10  sysBP            4240 non-null   float64
 11  diaBP            4240 non-null   float64
 12  BMI              4221 non-null   float64
 13  heartRate        4239 non-null   float64
 14  glucose          3852 non-null   float64
 15  TenYearCHD       4240 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 530.1 KB


The output of the `.info()` shows the following information:
- The dataset contains 4,240 patient records and 16 features, consuming a very lightweight 530.1 KB of memory.
- All 16 columns are already in a numerical format (integers or floats), meaning no text-to-number encoding is needed.
- 7 columns have missing values that require imputation. The glucose column has the highest number of missing entries (388), followed by education (105). The remaining nine columns have complete data. The missing values will be inspected in a latter section using the `.isnull().sum()` method.

#### **The `.shape` Method**

In [17]:
# Returns the shape of the data
df.shape

(4240, 16)

The output of the `.shape` method shows that the Framingham dataset contains 4,240 rows and 16 columns. This reinforces the output of the `.info()` method.

#### **Checking for Missing Values**

In [18]:
# Returns the number of missing values in each column
df.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

The output of the `.isnull().sum()` method reveals that 7 out of the 16 features contain missing values, totaling 645 missing entries across the dataset. The `glucose` feature has the highest concentration of missing data with 388 missing values, followed by `education` (105), `BPMeds` (53), `totChol` (50), `cigsPerDay` (29), `BMI` (19), and `heartRate` (1). The remaining 9 columns, including the target variable `TenYearCHD`, are fully complete with zero missing values.

### **Descriptive Statistics**

In [19]:
# Returns the summary statistics of the numerical columns in the dataset
df.describe()

,male,age,education,currentSmoker,cigsPerDay,BPMeds,prevalentStroke,prevalentHyp,diabetes,totChol,sysBP,diaBP,BMI,heartRate,glucose,TenYearCHD
count,4240.000000,4240.000000,4135.000000,4240.000000,4211.000000,4187.000000,4240.000000,4240.000000,4240.000000,4190.000000,4240.000000,4240.000000,4221.000000,4239.000000,3852.000000,4240.000000
mean,0.429245,49.580189,1.979444,0.494104,9.005937,0.029615,0.005896,0.310613,0.025708,236.699523,132.354599,82.897759,25.800801,75.878981,81.963655,0.151887
std,0.495027,8.572942,1.019791,0.500024,11.922462,0.169544,0.076569,0.462799,0.158280,44.591284,22.033300,11.910394,4.079840,12.025348,23.954335,0.358953
min,0.000000,32.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,107.000000,83.500000,48.000000,15.540000,44.000000,40.000000,0.000000
25%,0.000000,42.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,206.000000,117.000000,75.000000,23.070000,68.000000,71.000000,0.000000
50%,0.000000,49.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,234.000000,128.000000,82.000000,25.400000,75.000000,78.000000,0.000000
75%,1.000000,56.000000,3.000000,1.000000,20.000000,0.000000,0.000000,1.000000,0.000000,263.000000,144.000000,90.000000,28.040000,83.000000,87.000000,0.000000
max,1.000000,70.000000,4.000000,1.000000,70.000000,1.000000,1.000000,1.000000,1.000000,696.000000,295.000000,142.500000,56.800000,143.000000,394.000000,1.000000


The output of the `.describe()` method provides key insights into the central tendency, dispersion, and distribution shapes across the dataset's features:

* **Target Class Imbalance:** The target variable `TenYearCHD` has a mean of 0.1519, indicating that ~15.2% of participants experienced a 10-year coronary heart disease event. This confirms a significant class imbalance (~85% negative vs. ~15% positive), reinforcing the requirement for class rebalancing strategies during modeling.
* **Presence of Extreme Outliers:** Several continuous clinical variables exhibit substantial right-skewness, where the maximum value far exceeds the 75th percentile and median. Key examples include `totChol` (max: 696.0 vs. median: 234.0), `glucose` (max: 394.0 vs. median: 78.0), and `sysBP` (max: 295.0 vs. median: 128.0). This right-skewness strongly supports using median imputation rather than mean imputation.
* **Demographics & Behavioral Profiles:** The cohort ages range from 32 to 70 years old (mean: 49.6 years), with females representing ~57.1% of the dataset (`male` mean: 0.429). Approximately 49.4% are current smokers (`currentSmoker`), averaging ~9 cigarettes per day overall, with heavy smokers reporting up to 70 per day.
* **Clinical Conditions:** Prevalent hypertension (`prevalentHyp`) is common in the sample at 31.1%, whereas blood pressure medication usage (`BPMeds`: 3.0%), prior stroke (`prevalentStroke`: 0.6%), and diabetes (`diabetes`: 2.6%) represent low-prevalence clinical conditions.

### **Check for Duplicate Values**

In [21]:
# Returns the total number of duplicate rows in the dataset
df.duplicated().sum()

np.int64(0)

The output of the `.duplicated().sum()` method returns `0`, confirming that there are no duplicate rows across all 4,240 observations in the dataset. This verifies structural data integrity at the row level and confirms that no deduplication or record removal steps are required prior to preprocessing.

### **Target Variable Analysis: Class Distribution**

Before modeling, it is critical to evaluate the distribution of the target variable (`TenYearCHD`) to determine if there is a class imbalance. A severe imbalance can cause models to become biased toward the majority class.


In [22]:
# Returns the unique values and their counts in the target variable "TenYearCHD"
df["TenYearCHD"].value_counts()


TenYearCHD
0    3596
1     644
Name: count, dtype: int64

The output of the `.value_coiunts()` above confirms a significant class imbalance (3,596 negative cases vs. 644 positive cases). This indicates that the dataset is highly skewed toward patients without heart disease. Consequently, resampling techniques (such as SMOTE-ENN) and algorithm-level interventions (like Class Weights) must be implemented during model training to ensure the minority class is properly learned.

### **Target Variable Analysis: Feature Correlation**

Subsequently, the Pearson correlation coefficients between all independent physiological features and the target variable were calculated. This analysis was conducted to identify the primary linear risk factors for cardiovascular disease.

In [23]:
# Returns the correlation of all features with the target variable "TenYearCHD" and sorts them in descending order
df.corr()["TenYearCHD"].sort_values(ascending=False)

TenYearCHD         1.000000
age                0.225408
sysBP              0.216374
prevalentHyp       0.177458
diaBP              0.145112
glucose            0.125590
diabetes           0.097344
male               0.088374
BPMeds             0.087519
totChol            0.082369
BMI                0.075300
prevalentStroke    0.061823
cigsPerDay         0.057755
heartRate          0.022907
currentSmoker      0.019448
education         -0.054248
Name: TenYearCHD, dtype: float64

The features demonstrating the strongest positive linear correlation to heart disease risk are `age` (0.225), `sysBP` (0.216), and `prevalentHyp` (0.177). Because these primary clinical risk factors possess a clear linear relationship with the target variable, linear algorithms (such as Logistic Regression) are theoretically well-suited for this dataset. Consequently, it is hypothesized that linear models may perform competitively against complex non-linear ensembles (such as Random Forest or XGBoost).

**Next Phase:** Data Cleaning